### PR TITLE
SAAS-3138: Added ElastiCache Desired Node Type Plugin

### PR DESCRIPTION
**Pull request by sadeed12345**

### CODE REVIEW

File: plugins/aws/elasticache/elasticaheDesiredNodeType.js [hunk 1]
```diff
@@ -0,0 +1,72 @@
+var async = require('async');
+var helpers = require('../../../helpers/aws');
+
+module.exports = {
+    title: 'ElastiCache Desired Node Type',
+    category: 'ElastiCache',
+    domain: 'Databases',
+    description: 'Ensure that the Amazon ElastiCache cluster nodes provisioned in your AWS account have the desired node type established within your organization based on the workload deployed.',
+    more_info: 'Setting limits for the type of Amazon ElastiCache cluster nodes will help you address internal compliance requirements and prevent unexpected charges on your AWS bill.',
+    recommended_action: 'Enable Desired node type for ElastiCache clusters',
```
Comment (AkhtarAmir) [inline plugins/aws/elasticache/elasticaheDesiredNodeType.js:10]:
```suggestion
    recommended_action: 'Create ElastiCache clusters with desired node types',
```

File: plugins/aws/elasticache/elasticaheDesiredNodeType.js [hunk 2]
```diff
@@ -0,0 +1,72 @@
...
+    link: 'https://docs.aws.amazon.com/AmazonElastiCache/latest/red-ug/WhatIs.html',
+    apis: ['ElastiCache:describeCacheClusters'],
+    settings: {
+        elasticache_desired_node_type: {
```
Comment (AkhtarAmir) [inline plugins/aws/elasticache/elasticaheDesiredNodeType.js:14]: Create separate settings for redis and memcached clusters

File: collectors/aws/collector.js [hunk 3]
```diff
@@ -464,6 +464,12 @@ var calls = {
         }
 
     },
+    ElastiCache: {
```
Comment (AkhtarAmir) [inline collectors/aws/collector.js:467]: Merge master into this branch

**Review by AkhtarAmir — CHANGES_REQUESTED** [review]

File: plugins/aws/elasticache/elasticaheDesiredNodeType.js [hunk 4]
```diff
@@ -0,0 +1,71 @@
...
+           
+            for (var cluster of describeCacheClusters.data) {
+                if (!cluster.ARN) continue;
+
+                var resource = cluster.ARN;
+
+                if (cluster.CacheNodeType && cluster.CacheNodeType.length && cluster.Engine &&
+                    config.elasticache_desired_node_type.includes(cluster.CacheNodeType)) {
+                    helpers.addResult(results, 0,
+                        `cluster has desired node type: ${cluster.CacheNodeType} and engine: ${cluster.Engine} `,
```
Comment (AkhtarAmir) [inline plugins/aws/elasticache/elasticaheDesiredNodeType.js:58]:
```suggestion
                        `${cluster.Engine} Cluster has desired node type: ${cluster.CacheNodeType}`,
```

File: plugins/aws/elasticache/elasticaheDesiredNodeType.js [hunk 5]
```diff
@@ -0,0 +1,71 @@
...
+                var resource = cluster.ARN;
+
+                if (cluster.CacheNodeType && cluster.CacheNodeType.length && cluster.Engine &&
+                    config.elasticache_desired_node_type.includes(cluster.CacheNodeType)) {
+                    helpers.addResult(results, 0,
+                        `${cluster.Engine} Cluster has desired node type: ${cluster.CacheNodeType}`,
+                        region, resource);
+                } else {
+                    helpers.addResult(results, 2,
+                        `cluster does not have desired node type: ${cluster.CacheNodeType} and engine: ${cluster.Engine}`,
```
Comment (AkhtarAmir) [inline plugins/aws/elasticache/elasticaheDesiredNodeType.js:62]:
```suggestion
                        `${cluster.Engine} cluster does not have desired node type: ${cluster.CacheNodeType}`,
```

File: plugins/aws/elasticache/elasticaheDesiredNodeType.js [hunk 6]
```diff
@@ -0,0 +1,71 @@
...
+           
+            for (var cluster of describeCacheClusters.data) {
+                if (!cluster.ARN) continue;
+
+                var resource = cluster.ARN;
+
+                if (cluster.CacheNodeType && cluster.CacheNodeType.length && cluster.Engine &&
+                    config.elasticache_desired_node_type.includes(cluster.CacheNodeType)) {
+                    helpers.addResult(results, 0,
+                        `${cluster.Engine} Cluster has desired node type: ${cluster.CacheNodeType}`,
```
Comment (AkhtarAmir) [inline plugins/aws/elasticache/elasticaheDesiredNodeType.js:58]:
```suggestion
                        `${cluster.Engine} cluster has desired node type: ${cluster.CacheNodeType}`,
```

**Review by AkhtarAmir — APPROVED** [review]

File: plugins/aws/elasticache/elasticaheDesiredNodeType.js [hunk 7]
```diff
@@ -0,0 +1,71 @@
...
+                return rcb();
+            }
+
+            if (!describeCacheClusters.data.length) {
+                helpers.addResult(results, 0, 'No ElastiCache cluster found', region);
+                return rcb();
+            }
+           
+            for (var cluster of describeCacheClusters.data) {
+                if (!cluster.ARN) continue;
```
Comment (dipsubha06) [inline plugins/aws/elasticache/elasticaheDesiredNodeType.js:51]:
```suggestion
                if (!cluster.ARN || !cluster.Engine) continue;
```

**Pull request merged**